### PR TITLE
Moves CommandCompletionIT to Ledger API Test Tool

### LINKED ISSUE #2841: Port CommandCompletionServiceIT to the Ledger API Test Tool
Port `CommandCompletionServiceIT` to the Ledger API Test Tool.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Closes #2841

Also refactors test observers into a single API and unifies the existing implementations of the `DirectExecutionContext`.

As I was adding the tests, two yak shaves crossed my path:
- the number of `StreamObserver` implementations used for testing was getting out of hand
- we had two implementations of `DirectExecutionContext`

The bulk of the PR comes from addressing these two points.

If you feel they belong in their own PR I'm more than happy to suspend the work on this and break it into three chunks. Hopefully the PR should be relatively easy to review as the changes are bulky but relatively trivial.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
I had to exclude the new tests from running on top of the reference implementation of KVUtils, I'd leave this as is for now and open a ticket to make sure this is not an actual bug there. Other possible explanation are that we are testing for a guarantee that the Ledger API doesn't really provide and that the test is valid only for the Sandbox.

File: ledger/api-server-damlonx/reference-v2/BUILD.bazel [hunk 1]
```diff
@@ -137,6 +137,7 @@ server_conformance_test(
         "--exclude=LotsOfPartiesIT",
         "--exclude=TimeIT",
         "--exclude=ConfigManagementServiceIT",
+        "--exclude=CommandSubmissionCompletionIT",
```
Comment (gerolf-da) [inline ledger/api-server-damlonx/reference-v2/BUILD.bazel]: What's the reason for not running the test for the reference server?

**Comment by stefanobaghino-da** [comment]
As discussed offline:
- the tested behavior is not relevant for Ledger API implementations
- the behavior is relevant only for the sandbox
- the behavior is soon going to be made obsolete by the time model change

We therefore concluded that the test can be dropped.

File: ledger/api-server-damlonx/reference-v2/BUILD.bazel (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/api-server-damlonx/reference-v2/BUILD.bazel]: Addressed by https://github.com/digital-asset/daml/pull/3901/commits/75ab721d93ec36ae708c9c717ccda88c9a728164

The test does not represent a behavior that Ledger API implementations have to assume.

Furthermore, the behavior of checkpoints will soon be changed by the new time model.

We therefore decided to drop this test.

**Pull request merged**